Packer Task Fails Due to Temporary Directory Clashes

## Environment
Azure Pipelines on Windows 2012r2 private agent (version 2.144.2)

## Issue Description

On a machine which hosts multiple agents, there is an opportunity for conflicts with the temporary directory, which will cause the task to fail saying the directory already exists. For example:

```
Error: Unable to create directory 'C:\Users\REDACTED\AppData\Local\Temp\1552048259502'. EEXIST: file already exists, mkdir 'C:\Users\REDACTED\AppData\Local\Temp\1552048259502'
```

### Task logs

(Some information has been removed for privacy concerns.)

```
2019-03-08T12:30:58.1300601Z ##[debug]Evaluating condition for step: 'REDACTED'
2019-03-08T12:30:58.1301520Z ##[debug]Evaluating: succeeded()
2019-03-08T12:30:58.1301965Z ##[debug]Evaluating succeeded:
2019-03-08T12:30:58.1302515Z ##[debug]=> True
2019-03-08T12:30:58.1302990Z ##[debug]Result: True
2019-03-08T12:30:58.1303473Z ##[section]Starting: REDACTED
2019-03-08T12:30:58.1313311Z ==============================================================================
2019-03-08T12:30:58.1313650Z Task         : Build Machine Image
2019-03-08T12:30:58.1314286Z Description  : Build machine image using Packer. This image can be used for Azure Virtual machine scale set deployment
2019-03-08T12:30:58.1314626Z Version      : 0.0.19
2019-03-08T12:30:58.1314858Z Author       : Microsoft Corporation
2019-03-08T12:30:58.1315161Z Help         : [More Information](https://go.microsoft.com/fwlink/?linkid=845329)
2019-03-08T12:30:58.1315519Z ==============================================================================
2019-03-08T12:30:58.5431064Z ##[debug]agent.workFolder=f:\agent2\_work
2019-03-08T12:30:58.5498702Z ##[debug]loading inputs and endpoints
2019-03-08T12:30:58.5499337Z ##[debug]loading ENDPOINT_AUTH_PARAMETER_SYSTEMVSSCONNECTION_ACCESSTOKEN
2019-03-08T12:30:58.5509927Z ##[debug]loading ENDPOINT_AUTH_SCHEME_SYSTEMVSSCONNECTION
2019-03-08T12:30:58.5514342Z ##[debug]loading ENDPOINT_AUTH_SYSTEMVSSCONNECTION
2019-03-08T12:30:58.5517621Z ##[debug]loading INPUT_ADDITIONALBUILDERPARAMETERS
2019-03-08T12:30:58.5521595Z ##[debug]loading INPUT_BASEIMAGE
2019-03-08T12:30:58.5522127Z ##[debug]loading INPUT_BASEIMAGESOURCE
2019-03-08T12:30:58.5523299Z ##[debug]loading INPUT_CUSTOMIMAGEOSTYPE
2019-03-08T12:30:58.5529574Z ##[debug]loading INPUT_CUSTOMTEMPLATELOCATION
2019-03-08T12:30:58.5530214Z ##[debug]loading INPUT_CUSTOMTEMPLATEPARAMETERS
2019-03-08T12:30:58.5530891Z ##[debug]loading INPUT_PACKAGEPATH
2019-03-08T12:30:58.5531439Z ##[debug]loading INPUT_PACKERVERSION
2019-03-08T12:30:58.5532239Z ##[debug]loading INPUT_SKIPTEMPFILECLEANUPDURINGVMDEPROVISION
2019-03-08T12:30:58.5533651Z ##[debug]loading INPUT_TEMPLATETYPE
2019-03-08T12:30:58.9368093Z ##[debug]agent.proxyurl=undefined
2019-03-08T12:30:58.9369039Z ##[debug]VSTS_ARM_REST_IGNORE_SSL_ERRORS=undefined
2019-03-08T12:30:58.9369658Z ##[debug]AZURE_HTTP_USER_AGENT=VSTS_427430e9-64ea-4fae-aa50-7e52418cd9a8_release_396_56755_356632_1
2019-03-08T12:30:59.4780375Z ##[debug]Setting resource path to f:\agent2\_work\_tasks\PackerBuild_a99f71e8-625d-46a5-8ff9-94faa8cb14bf\0.0.20\task.json
2019-03-08T12:30:59.4781230Z ##[debug]check path : f:\agent2\_work\_tasks\PackerBuild_a99f71e8-625d-46a5-8ff9-94faa8cb14bf\0.0.20\task.json
2019-03-08T12:30:59.4786844Z ##[debug]set resource file to: f:\agent2\_work\_tasks\PackerBuild_a99f71e8-625d-46a5-8ff9-94faa8cb14bf\0.0.20\task.json
2019-03-08T12:30:59.4787340Z ##[debug]system.culture=en-US
2019-03-08T12:30:59.4898933Z ##[debug]templateType=custom
2019-03-08T12:30:59.4901743Z ##[debug]customTemplateLocation=f:\agent2\_work\r154\a\DevOps\packer\win2016.mo-application.json
2019-03-08T12:30:59.4902323Z ##[debug]check path : f:\agent2\_work\r154\a\DevOps\packer\win2016.mo-application.json
2019-03-08T12:30:59.5009316Z Parsing custom template parameters json.
2019-03-08T12:30:59.5012904Z ##[debug]customTemplateParameters={
 REDACTED
}
2019-03-08T12:30:59.5014099Z Parsing additional builder parameters json.
2019-03-08T12:30:59.5015013Z ##[debug]additionalBuilderParameters={}
2019-03-08T12:30:59.5022602Z ##[debug]skipTempFileCleanupDuringVMDeprovision=true
2019-03-08T12:30:59.5023167Z ##[debug]imageUri=null
2019-03-08T12:30:59.5023522Z ##[debug]packerVersion=1.3.2
2019-03-08T12:30:59.5027944Z ##[debug]which 'packer'
2019-03-08T12:30:59.5101406Z ##[debug]not found
2019-03-08T12:30:59.5101757Z Current installed packer version is null.
2019-03-08T12:30:59.5102319Z Either packer is not installed or its version is less than 1.3.2. Version 1.3.2 will be downloaded and installed for running task.
2019-03-08T12:30:59.5110682Z ##[debug]testing directory 'C:\Users\REDACTED\AppData\Local\Temp\1552048259502'
2019-03-08T12:30:59.5111247Z ##[debug]testing directory 'C:\Users\REDACTED\AppData\Local\Temp'
2019-03-08T12:30:59.5113046Z ##[debug]mkdir 'C:\Users\REDACTED\AppData\Local\Temp\1552048259502'
2019-03-08T12:30:59.5140196Z ##[debug]task result: Failed
2019-03-08T12:30:59.5195199Z ##[error]Error: Unable to create directory 'C:\Users\REDACTED\AppData\Local\Temp\1552048259502'. EEXIST: file already exists, mkdir 'C:\Users\REDACTED\AppData\Local\Temp\1552048259502'
2019-03-08T12:30:59.5205051Z ##[debug]Processed: ##vso[task.issue type=error;]Error: Unable to create directory 'C:\Users\REDACTED\AppData\Local\Temp\1552048259502'. EEXIST: file already exists, mkdir 'C:\Users\REDACTED\AppData\Local\Temp\1552048259502'
2019-03-08T12:30:59.5206611Z ##[debug]Processed: ##vso[task.complete result=Failed;]Error: Unable to create directory 'C:\Users\REDACTED\AppData\Local\Temp\1552048259502'. EEXIST: file already exists, mkdir 'C:\Users\REDACTED\AppData\Local\Temp\1552048259502'
2019-03-08T12:30:59.5248908Z ##[section]Finishing: REDACTED
```

### Error logs

```
2019-03-08T12:30:59.5110682Z ##[debug]testing directory 'C:\Users\REDACTED\AppData\Local\Temp\1552048259502'
2019-03-08T12:30:59.5111247Z ##[debug]testing directory 'C:\Users\REDACTED\AppData\Local\Temp'
2019-03-08T12:30:59.5113046Z ##[debug]mkdir 'C:\Users\REDACTED\AppData\Local\Temp\1552048259502'
2019-03-08T12:30:59.5140196Z ##[debug]task result: Failed
2019-03-08T12:30:59.5195199Z ##[error]Error: Unable to create directory 'C:\Users\REDACTED\AppData\Local\Temp\1552048259502'. EEXIST: file already exists, mkdir 'C:\Users\REDACTED\AppData\Local\Temp\1552048259502'
```

--- a/Tasks/PackerBuildV0/src/utilities.ts
+++ b/Tasks/PackerBuildV0/src/utilities.ts
@@ -107,7 +107,7 @@ export function findMatch(root: string, patterns: string[] | string): string[] {
 }
 
 export function getTempDirectory(): string {
-    return os.tmpdir();
+    return tl.getVariable('Agent.TempDirectory');
 }
 
 export function getCurrentTime(): number {

--- a/Tasks/PackerBuildV0/task.json
+++ b/Tasks/PackerBuildV0/task.json
@@ -15,7 +15,7 @@
     "version": {
         "Major": 0,
         "Minor": 0,
-        "Patch": 19
+        "Patch": 20
     },
     "demands": [],
     "minimumAgentVersion": "2.0.0",

--- a/Tasks/PackerBuildV0/task.loc.json
+++ b/Tasks/PackerBuildV0/task.loc.json
@@ -15,7 +15,7 @@
   "version": {
     "Major": 0,
     "Minor": 0,
-    "Patch": 19
+    "Patch": 20
   },
   "demands": [],
   "minimumAgentVersion": "2.0.0",

--- a/Tasks/PackerBuildV1/src/utilities.ts
+++ b/Tasks/PackerBuildV1/src/utilities.ts
@@ -107,7 +107,7 @@ export function findMatch(root: string, patterns: string[] | string): string[] {
 }
 
 export function getTempDirectory(): string {
-    return os.tmpdir();
+    return tl.getVariable('Agent.TempDirectory');
 }
 
 export function getCurrentTime(): number {

--- a/Tasks/PackerBuildV1/task.json
+++ b/Tasks/PackerBuildV1/task.json
@@ -15,7 +15,7 @@
     "version": {
         "Major": 1,
         "Minor": 0,
-        "Patch": 5
+        "Patch": 6
     },
     "demands": [],
     "preview": "true",

--- a/Tasks/PackerBuildV1/task.loc.json
+++ b/Tasks/PackerBuildV1/task.loc.json
@@ -15,7 +15,7 @@
   "version": {
     "Major": 1,
     "Minor": 0,
-    "Patch": 5
+    "Patch": 6
   },
   "demands": [],
   "preview": "true",

